content/v2.0/actions/parameters.md

```diff
@@ -0,0 +1,308 @@
+        end
+      end
+    end
+  end
+end
+```
+
+Even if `admin` is sent inside the body of the request, it isn't accessible from `params`.
+
+## Validations & Coercion
+
+### Use Cases
+
+In our example (called _"Signup"_), we want to make `password` a required param.
+
+Imagine we introduce a second feature: _"Invitations"_.
+An existing user can ask someone to join.
+Because the invitee will decide a password later on, we want to persist that `User` record without that value.
+
+If we put `password` validation in `User`, we need to handle these two use cases with a conditional.
+But in the long term this approach is painful from a maintenance perspective.
+
+```ruby
+# Example of poor style for validations
+class User
+  attribute :password, presence: { if: :password_required? }
+
+  private
+  
+  def password_required?
+     !invited_user? && !admin_password_reset?
+  end
+end
+```
+
+We can see validations as the set of rules for data correctness that we want for **a specific use case**.
+For us, a `User` can be persisted with or without a password, **depending on the workflow** and the route through
+which the `User` is persisted.
+
+### Boundaries
+
+The second important aspect is that we use validations to prevent invalid inputs to propagate in our system.
+In an MVC architecture, the model layer is the **farthest** from the input.
+It's expensive to check the data right before we create a record in the database.
+
+If we **consider correct data as a precondition** before starting our workflow, we should stop unacceptable inputs as soon as possible.
+
+Think of the following method.
+We don't want to continue if the data is invalid.
+
+```ruby
+def expensive_computation(argument)
+  return if argument.nil?
+  # ...
+end
+```
+
+### Usage
+
+We can coerce the Ruby type, validate if a param is required, determine if it is within a range of values, etc..
+
+```ruby
+# app/actions/signup/create.rb
+module Bookshelf
+  module Actions
+    module Signup
+      class Create < Action
+        MEGABYTE = 1024 ** 2
+
+        params do
+          required(:name).filled(:str?)
+          required(:email).filled(:str?, format?: /@/)
+          required(:password).filled(:str?)
+          required(:terms_of_service).filled(:bool?)
+          required(:age).filled(:int?, included_in?: 18..99)
+          optional(:avatar).filled(size?: 1..(MEGABYTE * 3))
+        end
+
+        def handle(req, res)
+          if req.params.valid?
+            # ...
+          else
+            # ...
+          end
+        end
+      end
+    end
+  end
+end
+```
+
+Parameter validations are delegated, under the hood, to [Hanami::Validations](https://github.com/hanami/validations).
+Please check the related documentation for a complete list of options and how to share code between validations.
+
+## Concrete Classes
+
+The params DSL is really quick and intuitive but it has the drawback that it can be visually noisy and makes it hard to unit test.
```

Comment:
This could be explained better. ie Why does it make things harder to unit test? I would also drop "unit test" and just say "test".